### PR TITLE
bpo-26175: Fix SpooledTemporaryFile IOBase abstract

### PR DESCRIPTION
**Pull request by GFernie**
One would assume that this class implements the IOBase abstract. As the
underlying file-like object is either io.BytesIO, io.StringIO, or a true
file object, this is a reasonable abstract expect and to implement.

Regardless, the behaviour of this class does not change much in the case
of the attribute being missing from the underlying file-like; an
AttributeError is still raised, albeit from one additional frame on the
stack trace.

https://bugs.python.org/issue26175

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

File: Lib/test/test_tempfile.py
Comment (merwok) [inline Lib/test/test_tempfile.py]: Minor: this test could be made simpler/faster with sets: assert that set(dir(f)) is a superset of set(dir(IOBase)) (or if that doesn’t work because IOBase has extra internal attributes, compare with `{'read', 'readable', etc}`).

Could you also add a test with isinstance to validate that the class implements IOBase?

**Review by merwok — CHANGES_REQUESTED** [review]
Patch looks good!  Note that the bug contains a stronger motivation than the first mssage here: «This was discovered when seeking a SpooledTemporaryFile-backed lzma file.»

**Comment by GFernie** [comment]
Thanks for the feedback @merwok 

I've made some changes to more accurately test how _SpooledTemporaryFile_ implements _IOBase_. So, there are now two separate tests:

- Assert that an instance of _SpooledTemporaryFile_ is an instance of _IOBase_. Because _IOBase_ already implements the interface (nearly) this makes it redundant to check for the individual attributes that _SpooledTemporaryFile_ should implement.
- Assert that the attributes which not declared in _IOBase_ but should be implemented by subclasses are implemented by _SpooledTemporaryFile_.

From the _IOBase_ docs: 

> Even though IOBase does not declare read(), readinto(), or write() because their signatures will vary, implementations and clients should consider those methods part of the interface.

I didn't expect the Spanish Inquisition!

**Review by merwok — APPROVED** [review]

**Comment by embe** [comment]
There is at least one more incompatibility with `IOBase:` `seek` should return the current position instead of `None`.

File: Lib/tempfile.py
Comment (pppery) [inline Lib/tempfile.py:682]: This shouldn't be added: deleting the SpooledTemporaryFile will null out the reference to the underline file, and therefor call its` __del__`

Comment (vadmium) [reply Lib/tempfile.py:682]: I think if you inherit the default `IOBase.__del__` implementation, it will call _close_ and defeat any ResourceWarning that might otherwise be emitted. Perhaps it is better to make `__del__` do nothing, ~~or set it to `object.__del__`~~. [Seems that last option doesn’t exist.]

**Comment by GFernie** [comment]
You are right, @embe; it's part of the interface:
https://docs.python.org/library/io.html#io.IOBase.seek

The method now returns the value returned from `self._file.__del__()`, delegating the responsibility of returning the current new absolute file position to the underlying file.

Thanks for pointing this out

File: Lib/tempfile.py
Comment (GFernie) [reply Lib/tempfile.py:682]: I agree: the underlying file should not be explicitly deleted as this is not expected behaviour. I can imagine a situation where someone is deliberately holding a reference to the underlying file and they would not expect/want it to be closed until their own reference has fallen out of scope. I've changed the method to do nothing and added a docstring to reflect this.

Thanks for your feedback

**Comment by GFernie** [comment]
Re seek: I've also done the same with `truncate`, as per Martin's recommendations on the Python bug tracker: https://bugs.python.org/issue26175#msg319145

This is also part of the file interface:
https://docs.python.org/library/io.html#io.IOBase.truncate

File: Lib/tempfile.py
Comment (pitrou) [reply Lib/tempfile.py:682]: > I agree: the underlying file should not be explicitly deleted as this is not expected behaviour

The doc says: """This function operates exactly as TemporaryFile() does, except [irrelevant differences]."""

And then, about TemporaryFile: """On completion of the context or destruction of the file object the temporary file will be removed from the filesystem."""

So it seems the underlying file _should_ be deleted when the file object disappears.

File: Lib/test/test_tempfile.py
Comment (pitrou) [inline Lib/test/test_tempfile.py:992]: Those tests are not very interesting. It would be better to test the methods operate properly.

**Review by pitrou — CHANGES_REQUESTED** [review]
Here are some comments. You will also need to update the documentation in `Doc/Library/tempfile.rst` (add a `versionchanged` tag documenting the change)

**Review by nubirstein — APPROVED** [review]
I had a case with python3.7 and latest Flask. When trying to send a file via Flask request, I got an error that SpooledTemporaryFile didn't have seekable() method. I ran my code with this tempfile.py implementation and it works perfectly. I hope to see these changes in the official python release very soon.

File: Lib/test/test_tempfile.py
Comment (coleifer) [reply Lib/test/test_tempfile.py:992]: It *literally* delegates everything to the underlying file.

The existing tests should cover the behaviors. The new behaviors are intended to flesh out the API so that `SpooledTemporaryFile`, and these tests seem to cover that.

File: Lib/tempfile.py
Comment (coleifer) [reply Lib/tempfile.py:682]: The point is that `IOBase` implements a `__del__` which has some side-effects. Those side effects are not desirable here. Any underlying buffers being wrapped, e.g. the `TemporaryFile` or the `BytesIO`, will be gc'd and handled as they should.

I think this implementation is correct.

**Comment by coleifer** [comment]
>  You will also need to update the documentation in Doc/Library/tempfile.rst (add a versionchanged tag documenting the change)

Something like:

```rst
.. versionchanged:: 3.19
   After a bunch of unhelpful bikeshedding, this class was unbroken and now behaves like you would expect.
```

**Comment by csabella** [comment]
@GFernie, please address the code review from @pitrou.  Thank you!

**Comment by csabella** [comment]
@GFernie hasn't had any activity on GitHub in over a year.  Does anyone else want to pick this up and address the outstanding code reviews?  If you do and you start with the existing PR, please credit GFernie is the initial commit message as a co-author.  Thanks!

File: Doc/library/tempfile.rst
Comment (merwok) [inline Doc/library/tempfile.rst]:
```suggestion
      Implemented :class:`io.IOBase` inteface.
```

simpler, no?

Comment (GFernie) [reply Doc/library/tempfile.rst]: Yes :slightly_smiling_face: thanks!

**Comment by GFernie** [comment]
@pitrou I've considered your comments on what we should expect when calling `__del__`. There's three options which have been discussed:

* Do nothing. The underlying buffer or temp file will be deleted when the object is finalised anyway. If the object is deleted with `del`, this would happen immediately. However, if `__del__` is called then this would delay the deletion of the underlying buffer or temp file until the object is deleted with `del` or is garbage collected, so possibly never. Here, we are in a situation where we called `__del__`, expected the resources to be freed, but not even a warning was raised.
* Inherit the `io.IOBase.__del__` implementation. The underlying buffer or temp file will be deleted when `__del__` is called, the object is deleted with `del`, or the object is garbage collected. This is the expected behaviour, but base implementation will explicitly call close, suppressing the _ResourceWarning_ which should be raised when the object is deleted before closing the file.
* Delegate the `__del__` call to the underlying file. If it is a buffer then then the buffer will be freed and if it's rolled over to disk then the temp file will be deleted. If the temp file was not closed before the object was deleted then a _ResourceWarning_ will be raised, just like _TemporaryFile_. This ensures that resources are freed exactly when we expect them to be and also that we are raising a warning when resources are not being managed properly by the programmer.

I've picked the third option, as you suggested. Thanks for your review.

I have made the requested changes; please review again.

File: Doc/library/tempfile.rst
Comment (pitrou) [inline Doc/library/tempfile.rst:119]: Since 3.8.0 has been released, this should be moved to a new `versionchanged:: 3.9` block.

File: Lib/tempfile.py
Comment (pitrou) [inline Lib/tempfile.py:683]: Hmm, calling the `__del__` method directly like that is not correct. `__del__` is only to be called when all external references are gone. Instead I would suggest something like the following:
```python
    def __del__(self):
        if not self.closed:
            import warnings
            warnings.warn('unclosed file %r' % (self,), ResourceWarning,
                          stacklevel=2, source=self)
            self.close()
```
(this is adapted from `FileIO.__del__` in `Lib/pyio.py`)

File: Lib/test/test_tempfile.py
Comment (pitrou) [inline Lib/test/test_tempfile.py:1025]: Does this actually work? Here I get something like:
```python
>>> f = tempfile.SpooledTemporaryFile()
>>> f.name
>>> f.rollover()
>>> f.name
3
```

File: Lib/test/test_tempfile.py
Comment (pitrou) [inline Lib/test/test_tempfile.py:1027]: Do we also expect a `ResourceWarning` to be raised when the file is not rolled over? If yes, there should be a test for it. If not, then the `__del__` implementation can be removed.

File: Lib/test/test_tempfile.py
Comment (pitrou) [inline Lib/test/test_tempfile.py:1028]: It is not correct to call `__del__` directly. Instead, use `del f` or `f = None`.

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for posting this. Generally speaking, calling `__del__` directly is incorrect. Only the interpreter should do that. From a user's point of view, what's important is whether an object is still reachable or not.

**Comment by csabella** [comment]
@GFernie, thank you for your last comments and for addressing the code review.  It looks like @pitrou had some additional notes after that, so it would be great if you could look at this again.  Thank you!

File: Lib/tempfile.py
Comment (thehesiod) [inline Lib/tempfile.py:759]: this else is not needed, PEP warning, might as well fix

Comment (merwok) [reply Lib/tempfile.py:759]: It's not necessary to remove the else.  It does no harm.

Comment (thehesiod) [reply Lib/tempfile.py:759]: I never said it was necessary, just bad style as it unnecessarily increased the code depth, take my comment as an FYI

**Comment by danieldjewell** [comment]
Not sure if @GFernie is still around -- are the outstanding changes/questions still those raised by @pitrou? @merwok?

File: Lib/tempfile.py
Comment (danieldjewell) [reply Lib/tempfile.py:759]: Just wanted to confirm @merwok - nothing needs to be changed here for approval?

Comment (merwok) [reply Lib/tempfile.py:759]: Right, I advise to not change the `else` line.

**Comment by merwok** [comment]
This PR needs to merge master and address Antoine’s latest review.

**Comment by csabella** [comment]
@danieldjewell, you're right.  Would you like to create a new PR based on this one?  If so, please follow this section of the [devguide](https://devguide.python.org/pullrequest/?highlight=original#converting-an-existing-patch-from-b-p-o-to-github) to give GFernie the credit as the original author.  Thanks!

**Comment by Molaire** [comment]
To Google people: this also breaks pandas.read_csv() and this monkey patch can be used for property readable writable and seekable to fix it.

> 
> 
> Hello,
> 
> For anyone that got here looking for the error `'SpooledTemporaryFile' object has no attribute 'seekable'`
> 
> In my case it was trying to pass an `SpooledTemporaryFile` to an `ZipFile` to decompress it (Starlette's UploadFile class if anyone cares). But there are other cases. Until the issue referenced in this PR is fixed, an option is to monkey patch the class. Sorry, ugly hack follows.
> 
> In one of your modules:
> 
> ```python
> from tempfile import SpooledTemporaryFile
> 
> def _seekable(self):
>     """
>     Monkey patched seekable() method for the SpooledTemporaryFile class.
> 
>     Sadly, we cannot send an instance of the SpooledTemporaryFile to the
>     ZipFile object.
> 
>     This is because the SpooledTemporaryFile doesn't inherit / implement the
>     IOBase class.
> 
>     This bug is reported in CPython:
> 
>         https://bugs.python.org/issue26175
> 
>     And an unmerged PR is located here:
> 
>         https://github.com/python/cpython/pull/3249/files
> 
>     In Python 3.8, attempting to pass the tmpfd fails with:
> 
>         AttributeError:
>             'SpooledTemporaryFile' object has no attribute 'seekable'
> 
>     So, the workaround to avoid copying the file is to monkey patch the class 
>     and add the missing method. Sorry.
>     """
>     return self._file.seekable()
> 
> 
> SpooledTemporaryFile.seekable = _seekable
> ```

**Comment by MikeEcho** [comment]
Python dev here from the future (2021 - save yourselves while you can). Still getting `SpooledTemporaryFile` errors while reading CSVs (via tempfiles) in. Kindly asking if requested changes will be addressed @GFernie ?

**Comment by pR0Ps** [comment]
@MikeEcho since it seemed like @GFernie had moved on I submitted an updated and rebased PR for this fix in #29560 with all of the code review comments addressed. Just waiting on a code review.

**Comment by methane** [comment]
Close this because I merged #29560.

**Pull request closed without merging**
